### ISSUE TITLE
Board searching

### ISSUE THREAD

**Issue by DonaldTsang** [DangerOnTheRanger/maniwani#59]
Basic board searching tools
1. search boards by name and partial names
2. Search board by descriptions and tags
3. Search board by categories (something good to implement)

**Comment by DangerOnTheRanger** [comment]
Searching would be a good feature to have; the infrastructure is more what needs to get worked out, I think. [Whoosh](https://whoosh.readthedocs.io/en/latest/index.html) would be my personal choice for the backend - it's portable, and pure Python, and doesn't have very high minimum system requirements. What I think should probably be discussed further are:
* How to handle indexes - Whoosh stores its indexes as flat files, which could potentially have to be something that we work around
* Communication - how would the searching/indexing part of Maniwani communicate with everything else? Is it its own separate Docker image, or integrated into the monolith?

**Comment by DonaldTsang** [comment]
A simple SQL table would do, you don't really need Whoosh as long as you know how to use string searching.

**Comment by DangerOnTheRanger** [comment]
You're right in that simply doing s SQL query over post subject/body would be a lot simpler, though it would be a bit more fragile as well in that SQL's LIKE does not fuzzy match, and fuzzy string searching is not part of the SQL standard (though some databases like Postgres do have fuzzy extensions), so you would be limited to searching for substrings and exact matches.

I'm largely okay with that limitation (most chans/archivers that I'm aware of are similarly limited), but I think it would be worth discussing more.

**Comment by DonaldTsang** [comment]
In that case we would need phonetic matching (https://8ch.net/hydrus/res/9068.html#q10232), word vectors (fasttext is a good one) or some other mechanism.

**Comment by DangerOnTheRanger** [comment]
fasttext might be worth a look since it seems at a glance not to take to much effort to add to the existing codebase, and at least as of late can run in an environment with a relatively small amount of memory, which is good. My gut feeling says to start out with using purely SQL LIKE and then build off of that if it's found to be insufficient, though. I'd rather not complicate things any more than necessary.

**Comment by DonaldTsang** [comment]
Please use phonetic matching as a first defence before you do word vectors, which is more computationally intensive.

**Comment by DangerOnTheRanger** [comment]
Good to know, I wasn't aware of that. [jellyfish](https://github.com/jamesturk/jellyfish) seems like a library that would be good for phonetic matching, so it could be worth a look.
